src/web4/handler.rs

```diff
@@ -0,0 +1,331 @@
+use near_sdk::base64::{
+    alphabet,
+    engine::{self, general_purpose},
+    Engine,
+};
+use serde_json::json;
+
+use crate::{
+    web4::types::{Web4Request, Web4Response},
+    Contract, Proposal,
+};
+
+pub const BASE64_ENGINE: engine::GeneralPurpose =
+    engine::GeneralPurpose::new(&alphabet::URL_SAFE, general_purpose::NO_PAD);
+
+pub fn web4_get(contract: &Contract, request: Web4Request) -> Web4Response {
+    let path_parts: Vec<&str> = request.path.split('/').collect();
+
+    let page = path_parts[1];
+    let mut title: String = String::from("near/dev/hub");
+    let mut description: String = String::from("The decentralized home base for NEAR builders");
+    let mut image: String = String::from(
+        "https://i.near.social/magic/large/https://near.social/magic/img/account/devhub.near",
+    );
+    let mut redirect_path: String = String::from("devhub.near/widget/app");
+
+    let mut initial_props_json = json!({"page": page}).to_string();
+
+    if path_parts.len() > 1 {
+        match page {
+            "community" => {
+                let handle = path_parts[2];
+                let community_option = contract.get_community(handle.to_string());
+                if community_option.is_some() {
+                    let community = community_option.unwrap();
+                    title = html_escape::encode_text(community.name.as_str()).to_string();
+                    description =
+                        html_escape::encode_text(community.description.as_str()).to_string();
+                    image = community.logo_url;
+                }
+                redirect_path = format!("devhub.near/widget/app?page={}&handle={}", page, handle);
+                initial_props_json = json!({"page": page, "handle": handle}).to_string();
+            }
+            "proposal" => {
+                let id_string = path_parts[2];
+                let id_option = id_string.parse::<u32>();
+                if id_option.is_ok() {
+                    let id = id_option.unwrap();
```

Comment:
This can be simplified to:

```suggestion
                if let Ok(id) = id_string.parse::<u32>() {
```